File: bit-manipulation/missing-number/missing-number.java
class Solution {
  public int missingNumber(int[] nums) {
    int n = nums.length;
    int missing = n;
    for (int i = 0; i < n; i++) {
      missing ^= i;
      missing ^= nums[i];
    }
    return missing;
  }

  public int missingNumber2(int[] nums) {
    int n = nums.length;
    int sum = n * (n + 1) / 2;

    for (int num : nums) {
      sum -= num;
    }

    return sum;
  }
}

File: bit-manipulation/missing-number/missing-number.go
package missingnumber

func missingNumber(nums []int) int {
	n := len(nums)
	missing := n

	for i := range nums {
		missing ^= i
		missing ^= nums[i]
	}
	return missing
}

func missingNumber2(nums []int) int {
	n := len(nums)
	sum := n * (n + 1) / 2

	for _, num := range nums {
		sum -= num
	}
	return sum
}
